SIGSEGV when attempting to remove both a service and associated check in anti-entropy

#### Overview of the Issue

Found the issue through a test flake. May or may not be easily reproducible in the wild.

Running `make test-flake FLAKE_PKG=command/connect/proxy FLAKE_TEST=TestRegisterMonitor FLAKE_N=100 FLAKE_CPUS=0.05` results in:

```
2020-01-22 17:40:48 - (1/100)
2020-01-22 17:40:58 - (2/100)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x17672bc]

goroutine 101 [running]:
github.com/hashicorp/consul/agent/local.(*State).deleteService(0xc0002ba0e0, 0xc0004c12c0, 0x9, 0x300010002, 0x1000200000003, 0xc0000bc720)
	/Users/chrispiraino/workspace/consul/agent/local/state.go:1100 +0x3ec
github.com/hashicorp/consul/agent/local.(*State).SyncChanges(0xc0002ba0e0, 0x0, 0x0)
	/Users/chrispiraino/workspace/consul/agent/local/state.go:1037 +0x202
github.com/hashicorp/consul/agent/local.(*State).SyncFull(0xc0002ba0e0, 0x0, 0x308981f)
	/Users/chrispiraino/workspace/consul/agent/local/state.go:1018 +0x5c
github.com/hashicorp/consul/agent/ae.(*StateSyncer).nextFSMState(0xc00004cf80, 0x307e532, 0x8, 0x307e532, 0x8)
make: *** [test-flake] Error 1
```

Investigating further, the seg fault appears to happen because of attempting to delete/prune a after a `SyncFull` call, which ends up re-adding the `CheckID` to the `checks` map [here](https://github.com/hashicorp/consul/blob/b71630b752539068e48f44367e1179ad55685864/agent/local/state.go#L962), That is then referenced [here](https://github.com/hashicorp/consul/blob/b71630b752539068e48f44367e1179ad55685864/agent/local/state.go#L1096), catching a `nil` value of `c.Check` and seg faulting.

Wrote up a better test reproduction and will add it as a unit test in the fix.

--- a/agent/local/state.go
+++ b/agent/local/state.go
@@ -70,7 +70,7 @@ type CheckState struct {
 	// Check is the local copy of the health check record.
 	//
 	// Must Clone() the overall CheckState before mutating this. After mutation
-	// reinstall into the checks map.
+	// reinstall into the checks map. If Deleted is true, this field can be nil.
 	Check *structs.HealthCheck
 
 	// Token is the ACL record to update or delete the health check

--- a/agent/local/state.go
+++ b/agent/local/state.go
@@ -1093,7 +1093,7 @@ func (l *State) deleteService(key structs.ServiceID) error {
 		delete(l.services, key)
 		// service deregister also deletes associated checks
 		for _, c := range l.checks {
-			if c.Deleted && c.Check.ServiceID == key.ID {
+			if c.Deleted && c.Check != nil && c.Check.ServiceID == key.ID {
 				l.pruneCheck(c.Check.CompoundCheckID())
 			}
 		}

--- a/agent/local/state_test.go
+++ b/agent/local/state_test.go
@@ -1081,6 +1081,81 @@ func TestAgentAntiEntropy_Checks(t *testing.T) {
 	}
 }
 
+func TestAgentAntiEntropy_RemovingServiceAndCheck(t *testing.T) {
+	t.Parallel()
+	a := agent.NewTestAgent(t, t.Name(), "")
+	defer a.Shutdown()
+
+	testrpc.WaitForTestAgent(t, a.RPC, "dc1")
+	// Register info
+	args := &structs.RegisterRequest{
+		Datacenter: "dc1",
+		Node:       a.Config.NodeName,
+		Address:    "127.0.0.1",
+	}
+
+	var out struct{}
+
+	// Exists remote (delete)
+	svcID := "deleted-check-service"
+	srv := &structs.NodeService{
+		ID:      svcID,
+		Service: "echo",
+		Tags:    []string{},
+		Address: "127.0.0.1",
+		Port:    8080,
+	}
+	args.Service = srv
+	if err := a.RPC("Catalog.Register", args, &out); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Exists remote (delete)
+	chk := &structs.HealthCheck{
+		Node:           a.Config.NodeName,
+		CheckID:        "lb",
+		Name:           "lb",
+		ServiceID:      svcID,
+		Status:         api.HealthPassing,
+		EnterpriseMeta: *structs.DefaultEnterpriseMeta(),
+	}
+
+	args.Check = chk
+	if err := a.RPC("Catalog.Register", args, &out); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	if err := a.State.SyncFull(); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	var services structs.IndexedNodeServices
+	req := structs.NodeSpecificRequest{
+		Datacenter: "dc1",
+		Node:       a.Config.NodeName,
+	}
+
+	if err := a.RPC("Catalog.NodeServices", &req, &services); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// The consul service will still be registered
+	if len(services.NodeServices.Services) != 1 {
+		t.Fatalf("Expected all services to be deleted, got: %#v", services.NodeServices.Services)
+	}
+
+	var checks structs.IndexedHealthChecks
+	// Verify that we are in sync
+	if err := a.RPC("Health.NodeChecks", &req, &checks); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// The serfHealth check will still be here
+	if len(checks.HealthChecks) != 1 {
+		t.Fatalf("Expected the health check to be deleted, got: %#v", checks.HealthChecks)
+	}
+}
+
 func TestAgentAntiEntropy_Checks_ACLDeny(t *testing.T) {
 	t.Parallel()
 	dc := "dc1"

